File: app/src/main/java/com/datarockets/mnchkn/store/GameServiceImpl.java
package com.datarockets.mnchkn.store;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.preference.PreferenceManager;

import com.datarockets.mnchkn.models.GameStep;
import com.datarockets.mnchkn.models.Player;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import lecho.lib.hellocharts.model.Line;
import lecho.lib.hellocharts.model.LineChartData;
import lecho.lib.hellocharts.model.PointValue;

public class GameServiceImpl implements GameService {

    private static final String IS_GAME_STARTED = "is_game_started";

    private MunchkinDatabaseHelper mDatabase;
    private SharedPreferences mPreferences;
    private SharedPreferences.Editor mPreferencesEditor;
    private Map<Player, List<GameStep>> mGameStepsMap;

    public GameServiceImpl(Context context) {
        mDatabase = new MunchkinDatabaseHelper(context);
        mPreferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Override
    public void insertStep(Player player) {
        GameStep gameStep = new GameStep();
        gameStep.setPlayerId(player.getId());
        gameStep.setPlayerLevel(player.getLevelScore());
        gameStep.setPlayerStrength(player.getStrengthScore());
        mDatabase.insertStep(gameStep);
    }

    @Override
    public void createPlayerIdGameStepsMap() {
        List<Player> playersList = mDatabase.getPlayers();
        List<GameStep> gameSteps = mDatabase.getGameSteps();

        Map<Player, List<GameStep>> playerGameStepsMap = new HashMap<>();
        for (Player player : playersList) {
            List<GameStep> playerSteps = new ArrayList<>();
            for (GameStep step : gameSteps) {
                if (step.getPlayerId() == player.getId()) {
                    playerSteps.add(step);
                }
            }
            playerGameStepsMap.put(player, playerSteps);
        }
        mGameStepsMap = playerGameStepsMap;
    }

    @Override
    public Map<Player, List<GameStep>> getScoresChartData() {
        return mGameStepsMap;
    }

    @Override
    public LineChartData createScoresChartData(int type, Map<Player, List<GameStep>> gameStepsMap) {
        List<Line> playersLines = new ArrayList<>();
        List<String> playerColors = new ArrayList<>();
        LineChartData lineChartData;

        for (Player player : gameStepsMap.keySet()) {
            String color = player.getColor();
            playerColors.add(color);
        }

        for (List<GameStep> gameSteps : gameStepsMap.values()) {
            List<PointValue> pointValues = new ArrayList<>();
            for (int i = 0; i < gameSteps.size(); i++) {
                switch (type) {
                    case 0:
                        int playerLevel = gameSteps.get(i).getPlayerLevel();
                        pointValues.add(new PointValue(i, playerLevel));
                        break;
                    case 1:
                        int playerStrength = gameSteps.get(i).getPlayerStrength();
                        pointValues.add(new PointValue(i, playerStrength));
                        break;
                    case 2:
                        int playerTotal = gameSteps.get(i).getPlayerLevel() +
                                gameSteps.get(i).getPlayerStrength();
                        pointValues.add(new PointValue(i, playerTotal));
                        break;
                }
            }
            Line line = new Line();
            line.setValues(pointValues);
            line.setHasPoints(false);
            line.setCubic(true);
            playersLines.add(line);
        }

        for (int i = 0; i < playersLines.size(); i++) {
            playersLines.get(i).setColor(Color.parseColor(playerColors.get(i)));
        }

        lineChartData = new LineChartData(playersLines);

        return lineChartData;
    }

    @Override
    public void clearSteps() {
        mDatabase.clearSteps();
    }

    @Override
    public void setGameStatus(boolean status) {
        mPreferencesEditor = mPreferences.edit();
        mPreferencesEditor.putBoolean(IS_GAME_STARTED, status);
        mPreferencesEditor.apply();
    }

    public boolean isGameStarted() {
        return mPreferences.getBoolean(IS_GAME_STARTED, false);
    }

}

File: app/src/main/kotlin/com/datarockets/mnchkn/store/GameServiceImpl.kt
package com.datarockets.mnchkn.store

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Color
import com.datarockets.mnchkn.MunchkinApplication
import com.datarockets.mnchkn.models.GameStep
import com.datarockets.mnchkn.models.Player
import lecho.lib.hellocharts.model.Line
import lecho.lib.hellocharts.model.LineChartData
import lecho.lib.hellocharts.model.PointValue
import java.util.*
import javax.inject.Inject

class GameServiceImpl(context: Context) : GameService {

    @Inject
    lateinit var mDatabase: MunchkinDatabaseHelper
    @Inject
    lateinit var mPreferences: SharedPreferences
    @Inject
    lateinit var mPreferencesEditor: SharedPreferences.Editor

    override var scoresChartData = mutableMapOf<Player, List<GameStep>>()
        private set

    init {
        MunchkinApplication.get(context).appComponent.inject(this)
    }

    override fun insertStep(player: Player) {
        val gameStep = GameStep(
                playerId = player.id,
                playerLevel = player.levelScore,
                playerStrength = player.strengthScore
        )
        mDatabase.insertStep(gameStep)
    }

    override fun createPlayerIdGameStepsMap() {
        val playersList = mDatabase.players
        val gameSteps = mDatabase.gameSteps

        val playerGameStepsMap = HashMap<Player, List<GameStep>>()
        for (player in playersList) {
            val playerSteps = ArrayList<GameStep>()
            for (step in gameSteps) {
                if (step.playerId == player.id) {
                    playerSteps.add(step)
                }
            }
            playerGameStepsMap.put(player, playerSteps)
        }
        scoresChartData = playerGameStepsMap
    }

    override fun createScoresChartData(type: Int, gameStepsMap: Map<Player, List<GameStep>>): LineChartData {
        val playersLines = ArrayList<Line>()
        val playerColors = ArrayList<String>()
        val lineChartData: LineChartData

        for (player in gameStepsMap.keys) {
            val color = player.color
            playerColors.add(color)
        }

        for (gameSteps in gameStepsMap.values) {
            val pointValues = ArrayList<PointValue>()
            for (i in gameSteps.indices) {
                when (type) {
                    0 -> {
                        val playerLevel = gameSteps[i].playerLevel
                        pointValues.add(PointValue(i.toFloat(), playerLevel.toFloat()))
                    }
                    1 -> {
                        val playerStrength = gameSteps[i].playerStrength
                        pointValues.add(PointValue(i.toFloat(), playerStrength.toFloat()))
                    }
                    2 -> {
                        val playerTotal = gameSteps[i].playerLevel + gameSteps[i].playerStrength
                        pointValues.add(PointValue(i.toFloat(), playerTotal.toFloat()))
                    }
                }
            }
            val line = Line()
            line.values = pointValues
            line.setHasPoints(false)
            line.isCubic = true
            playersLines.add(line)
        }

        for (i in playersLines.indices) {
            playersLines[i].color = Color.parseColor(playerColors[i])
        }

        lineChartData = LineChartData(playersLines)

        return lineChartData
    }

    override fun clearSteps() {
        mDatabase.clearSteps()
    }

    override fun setGameStatus(status: Boolean) {
        mPreferencesEditor = mPreferences.edit()
        mPreferencesEditor.putBoolean(IS_GAME_STARTED, status)
        mPreferencesEditor.apply()
    }

    override val isGameStarted: Boolean
        get() = mPreferences.getBoolean(IS_GAME_STARTED, false)

    companion object {

        private val IS_GAME_STARTED = "is_game_started"
    }

}
